### PR TITLE
feat: tracing support for timer calls

### PR DESCRIPTION
**Pull request by aludwiko**
This is how it looks like: 
image

### CODE REVIEW

File: samples/java-spring-choreography-saga-quickstart/src/it/java/user/registry/UserCreationIntegrationTest.java [hunk 1]
```diff
@@ -108,10 +108,7 @@ public void testUserCreationFailureDueToInvalidInput() throws Exception {
 
     await()
       .ignoreExceptions()
-      // timer will fire in 3 seconds and un-reserve the email
-      // see it/resources/application.conf for the configuration
-      // we only start to polling after 3 seconds to give the timer a chance to fire
-      .between(Duration.ofSeconds(3), Duration.ofSeconds(6))
+      .timeout(Duration.ofSeconds(10)) //3 seconds for the projection lag + 3 seconds for the timer to fire
```
Comment (aludwiko) [inline samples/java-spring-choreography-saga-quickstart/src/it/java/user/registry/UserCreationIntegrationTest.java:111]: @octonato for the record, using `between` has different semantics, it starts pooling immediately, but the assertion must be valid only within the boundaries.

**Comment by franciscolopezsancho** [comment]
From what I understand:
We add metadata to the TimerScheduler so we can pass it to the TimerScheduler when is created in Workflow or Action.  This way when starting a timer (startSingleTimer) or cancel it because the metadata is available and we can add it as a header.  

With all of this later when the timer calls the component, it already has in the header the trace parent and so the Kalix Runtime registers a span as the child of that parent. 

LGTM 

`/kalix.timers.TimersService/AddSingle` it's the gRPC call to the timer to `startSingleTimer`, right? Or maybe the timer calling the component?
What's the reason not registering the call to the component that the timer will(did) make?

**Comment by aludwiko** [comment]
> With all of this later when the timer calls the component, it already has in the header the trace parent and so the Kalix Runtime registers a span as the child of that parent.

Nope, when a timer calls a component, it's a new and separate trace.

> /kalix.timers.TimersService/AddSingle it's the gRPC call to the timer to startSingleTimer, right? Or maybe the timer calling the component?

it's call to schedule a new timer.

> What's the reason not registering the call to the component that the timer will(did) make? I saw you mention something yesterday in Slack.

would it make sense to have a trace that takes (possibly) months?

File: sdk/java-sdk-protobuf/src/main/scala/kalix/javasdk/impl/timer/TimerSchedulerImpl.scala [hunk 2]
```diff
@@ -64,13 +75,25 @@ private[kalix] final class TimerSchedulerImpl(val messageCodec: MessageCodec, va
...
+
+  private def addHeaders[I, O](
+      callBuilder: SingleResponseRequestBuilder[I, O],
+      metadata: Metadata): SingleResponseRequestBuilder[I, O] = {
+    metadata.asScala.foldLeft(callBuilder) { case (builder, entry) =>
+      if (entry.isText) {
+        builder.addHeader(entry.getKey, entry.getValue)
+      } else {
+        builder
+      }
```
[the remark this answers is not in the document]
Comment (franciscolopezsancho) [reply sdk/java-sdk-protobuf/src/main/scala/kalix/javasdk/impl/timer/TimerSchedulerImpl.scala:95]: Personal preference, feel free to skip it.

**Comment by franciscolopezsancho** [comment]
> it's call to schedule a new timer.

That explain both questions 👍 

> would it make sense to have a trace that takes (possibly) months?

I'd say that's very unlikely so majority of cases would be interesting I feel. But that might be another PR in case we agree. Another related question: how about adding in this current `AddSingle` span some info about what will be called in the future?

**Review by franciscolopezsancho — APPROVED** [review]
LGTM

**Comment by franciscolopezsancho** [comment]
Sorry my bad in one suggestion. One sec

**Comment by aludwiko** [comment]
> Another related question: how about adding in this current AddSingle span some info about what will be called in the future? 

I am not sure we can do this, but it can be added later.

**Comment by franciscolopezsancho** [comment]
> I am not sure we can do this, but it can be added later.

agree

**Comment by franciscolopezsancho** [comment]
https://github.com/lightbend/kalix-jvm-sdk/issues/2038

**Pull request merged**